File: src/tools/pi.ts
import type { AgentRequest, AgentResult, AgentUsage } from "./types.js";
import { spawnAgent } from "./spawn.js";

/** Pi accepts every qc_thinking value as `--thinking`. */
export function buildPiArgs(request: AgentRequest): string[] {
  const args = ["--mode", "json", "-a", "--session-id", request.sessionId];
  if (request.model) args.push("--model", request.model);
  if (request.thinking) args.push("--thinking", request.thinking);
  if (request.noSkills) args.push("--no-skills");
  if (request.skillPath) args.push("--skill", request.skillPath);
  return args;
}

function textPartsFromContent(content: unknown): string[] {
  if (typeof content === "string") return [content];
  if (!Array.isArray(content)) return [];
  const texts: string[] = [];
  for (const part of content) {
    if (part && typeof part === "object" && (part as { type?: string }).type === "text" && typeof (part as { text?: string }).text === "string") {
      texts.push((part as { text: string }).text);
    }
  }
  return texts;
}

function usageFromPi(raw: unknown): AgentUsage | undefined {
  if (!raw || typeof raw !== "object") return undefined;
  const u = raw as Record<string, unknown>;
  const out: AgentUsage = {};
  if (typeof u.input === "number") out.input_tokens = u.input;
  if (typeof u.output === "number") out.output_tokens = u.output;
  if (typeof u.cacheRead === "number") out.cache_read_tokens = u.cacheRead;
  if (typeof u.reasoning === "number") out.thinking_tokens = u.reasoning;
  if (typeof u.totalTokens === "number") out.total_tokens = u.totalTokens;
  const cost = u.cost;
  if (cost && typeof cost === "object" && typeof (cost as { total?: unknown }).total === "number") {
    out.cost = (cost as { total: number }).total;
  }
  return Object.keys(out).length ? out : undefined;
}

function eventUsage(event: Record<string, unknown>): unknown {
  if (event.usage !== undefined) return event.usage;
  const message = event.message;
  if (message && typeof message === "object" && (message as { usage?: unknown }).usage !== undefined) {
    return (message as { usage: unknown }).usage;
  }
  return undefined;
}

/**
 * Extract assistant text and session id from Pi JSONL.
 * Live `--mode json` writes the reply on assistant `message_end.message.content[].text`.
 * User `message_end` echoes the prompt; skip it. Skip thinking parts (keep `type:text` only).
 * Session files (and older streams) use `{type:"message", role:"assistant"}` — keep as fallback.
 */
export function parsePiOutput(stdout: string, fallbackSessionId: string): { text: string; nativeId: string; result: unknown; error?: string; usage?: AgentUsage } {
  const lines = stdout
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);
  let nativeId = fallbackSessionId;
  const liveTexts: string[] = [];
  const fallbackTexts: string[] = [];
  const events: unknown[] = [];
  let error: string | undefined;
  let endUsage: unknown;
  let updateUsage: unknown;

  for (const line of lines) {
    let event: Record<string, unknown>;
    try {
      event = JSON.parse(line) as Record<string, unknown>;
    } catch {
      continue;
    }
    events.push(event);
    if (event.type === "session" && typeof event.id === "string" && event.id) {
      nativeId = event.id;
    }
    if (event.type === "message_update") {
      const raw = eventUsage(event);
      if (raw !== undefined) updateUsage = raw;
    }
    if (event.type === "message_end") {
      const message = event.message;
      // Live JSONL echoes the user turn as message_end too; only assistant text is the reply.
      if (message && typeof message === "object") {
        const msg = message as { role?: string; content?: unknown; stopReason?: string; errorMessage?: string; usage?: unknown };
        if (msg.role === "assistant") {
          liveTexts.push(...textPartsFromContent(msg.content));
          if (msg.usage !== undefined) endUsage = msg.usage;
        }
        if (msg.stopReason === "error") {
          error = typeof msg.errorMessage === "string" && msg.errorMessage.trim() ? msg.errorMessage.trim() : "pi reported a run error";
        }
      }
    }
    if (event.type === "message" && event.role === "assistant") {
      fallbackTexts.push(...textPartsFromContent(event.content));
    }
    // Some Pi builds emit assistant text as a dedicated text event.
    if (event.type === "text" && typeof event.text === "string") fallbackTexts.push(event.text);
  }

  const usage = usageFromPi(endUsage ?? updateUsage);
  return {
    text: (liveTexts.length ? liveTexts : fallbackTexts).join(""),
    nativeId,
    result: events,
    ...(error ? { error } : {}),
    ...(usage ? { usage } : {}),
  };
}

export async function runPi(request: AgentRequest): Promise<AgentResult> {
  const args = buildPiArgs(request);
  const capture = await spawnAgent(request.command, args, request.prompt, request.workdir, request.env, "pi");
  const parsed = parsePiOutput(capture.stdout, request.sessionId);
  return {
    text: parsed.text,
    nativeId: parsed.nativeId,
    exit: capture.exit,
    result: parsed.result,
    stderr: capture.stderr,
    argv: args,
    stdoutBytes: capture.stdout.length,
    ...(parsed.error ? { error: parsed.error } : {}),
    ...(parsed.usage ? { usage: parsed.usage } : {}),
  };
}

File: src/tools/pi.rs
use serde_json::Value;

use crate::errors::QcError;
use crate::tools::spawn::spawn_agent;
use crate::tools::types::{AgentRequest, AgentResult, AgentUsage, utf16_units};

pub fn build_pi_args(request: &AgentRequest) -> Vec<String> {
    let mut args = vec![
        "--mode".into(),
        "json".into(),
        "-a".into(),
        "--session-id".into(),
        request.session_id.clone(),
    ];
    if let Some(model) = &request.model {
        args.push("--model".into());
        args.push(model.clone());
    }
    if let Some(thinking) = &request.thinking {
        args.push("--thinking".into());
        args.push(thinking.clone());
    }
    if request.no_skills == Some(true) {
        args.push("--no-skills".into());
    }
    if let Some(skill) = &request.skill_path {
        args.push("--skill".into());
        args.push(skill.clone());
    }
    args
}

fn text_parts_from_content(content: &Value) -> Vec<String> {
    if let Some(text) = content.as_str() {
        return vec![text.to_string()];
    }
    let Some(items) = content.as_array() else {
        return Vec::new();
    };
    let mut texts = Vec::new();
    for part in items {
        if part.get("type").and_then(Value::as_str) == Some("text")
            && let Some(text) = part.get("text").and_then(Value::as_str)
        {
            texts.push(text.to_string());
        }
    }
    texts
}

fn usage_from_pi(raw: Option<&Value>) -> Option<AgentUsage> {
    let u = raw.and_then(Value::as_object)?;
    let mut out = AgentUsage::default();
    if let Some(n) = u.get("input").and_then(Value::as_f64) {
        out.input_tokens = Some(n);
    }
    if let Some(n) = u.get("output").and_then(Value::as_f64) {
        out.output_tokens = Some(n);
    }
    if let Some(n) = u.get("cacheRead").and_then(Value::as_f64) {
        out.cache_read_tokens = Some(n);
    }
    if let Some(n) = u.get("reasoning").and_then(Value::as_f64) {
        out.thinking_tokens = Some(n);
    }
    if let Some(n) = u.get("totalTokens").and_then(Value::as_f64) {
        out.total_tokens = Some(n);
    }
    if let Some(n) = u
        .get("cost")
        .and_then(Value::as_object)
        .and_then(|c| c.get("total"))
        .and_then(Value::as_f64)
    {
        out.cost = Some(n);
    }
    if out.is_empty() { None } else { Some(out) }
}

fn event_usage(event: &Value) -> Option<&Value> {
    if event.get("usage").is_some() {
        return event.get("usage");
    }
    event.get("message").and_then(|m| m.get("usage"))
}

pub struct ParsedPi {
    pub text: String,
    pub native_id: String,
    pub result: Value,
    pub error: Option<String>,
    pub usage: Option<AgentUsage>,
}

/// Live `--mode json` assistant `message_end`; legacy message/text fallback.
pub fn parse_pi_output(stdout: &str, fallback_session_id: &str) -> ParsedPi {
    let mut native_id = fallback_session_id.to_string();
    let mut live_texts = Vec::new();
    let mut fallback_texts = Vec::new();
    let mut events = Vec::new();
    let mut error = None;
    let mut end_usage = None;
    let mut update_usage = None;

    for line in stdout.lines().map(str::trim).filter(|l| !l.is_empty()) {
        let Ok(event) = serde_json::from_str::<Value>(line) else {
            continue;
        };
        events.push(event.clone());
        if event.get("type").and_then(Value::as_str) == Some("session")
            && let Some(id) = event
                .get("id")
                .and_then(Value::as_str)
                .filter(|s| !s.is_empty())
        {
            native_id = id.to_string();
        }
        if event.get("type").and_then(Value::as_str) == Some("message_update")
            && let Some(raw) = event_usage(&event)
        {
            update_usage = Some(raw.clone());
        }
        if event.get("type").and_then(Value::as_str) == Some("message_end")
            && let Some(message) = event.get("message")
        {
            if message.get("role").and_then(Value::as_str) == Some("assistant") {
                if let Some(content) = message.get("content") {
                    live_texts.extend(text_parts_from_content(content));
                }
                if let Some(usage) = message.get("usage") {
                    end_usage = Some(usage.clone());
                }
            }
            if message.get("stopReason").and_then(Value::as_str) == Some("error") {
                let detail = message
                    .get("errorMessage")
                    .and_then(Value::as_str)
                    .map(str::trim)
                    .filter(|s| !s.is_empty())
                    .unwrap_or("pi reported a run error");
                error = Some(detail.to_string());
            }
        }
        if event.get("type").and_then(Value::as_str) == Some("message")
            && event.get("role").and_then(Value::as_str) == Some("assistant")
            && let Some(content) = event.get("content")
        {
            fallback_texts.extend(text_parts_from_content(content));
        }
        if event.get("type").and_then(Value::as_str) == Some("text")
            && let Some(text) = event.get("text").and_then(Value::as_str)
        {
            fallback_texts.push(text.to_string());
        }
    }

    let text = if live_texts.is_empty() {
        fallback_texts.join("")
    } else {
        live_texts.join("")
    };
    ParsedPi {
        text,
        native_id,
        result: Value::Array(events),
        error,
        usage: usage_from_pi(end_usage.as_ref().or(update_usage.as_ref())),
    }
}

pub fn run_pi(request: &AgentRequest) -> Result<AgentResult, QcError> {
    let args = build_pi_args(request);
    let capture = spawn_agent(
        &request.command,
        &args,
        Some(&request.prompt),
        &request.workdir,
        &request.env,
        "pi",
    )?;
    let parsed = parse_pi_output(&capture.stdout, &request.session_id);
    Ok(AgentResult {
        text: parsed.text,
        native_id: parsed.native_id,
        exit: capture.exit,
        result: parsed.result,
        stderr: capture.stderr,
        argv: args,
        stdout_bytes: utf16_units(&capture.stdout),
        error: parsed.error,
        usage: parsed.usage,
    })
}
